text/0000-core-builder.md

```diff
@@ -0,0 +1,46 @@
+# Ubuntu Core Builder
+
+## Summary
+
+Over the past, the Paketo project has created four different stacks and builders based on the Ubuntu Linux distribution. The difference is in the number of packages included which in turn is based on the requirements of the different buildpacks included.
+This RFC proposes to create a new stack repository that will create a single `build` image and a number of `run` images based on the Ubuntu distribution and replace the existing `static`, `tiny` and `base` stacks.
+This RFC further proposes to create a new builder repository that will create a single builder using the build image as a base layer and the `base` run image as the default `run` image.
+
+Users will be able to use the new builder to build and run their applications using any of the language family buildpacks included. Instead of requireing a decision for an optimised builder, they can "just" start and later decide on an optimised `run` image.
+
+## Motivation
+
+The Paketo project will be easier to use for users - they can start without being asked to decide on optimisation.
+The stacks and builders maintainer team will have less work to do - they will only need to maintain a single stack and builder instead of three.
+
+## Detailed Explanation
+
+We would replace the current `base`, `tiny` and `static` stacks with a single stack that includes all packages required by the buildpacks. Similarly, we would replace the current `base`, `tiny` and `static` builders with a single builder.
```

Comment:
One thing that I think we should write into the detailed explanation is that we will by default squash layers in the builder. I propose that we do this per composite image. So our builder will contain *only* composite images and each composite image will get squashed into one layer, so that our builder does not exceed the layer limit. The pack cli supports this now, so it should be feasible.

Reply:
This seems to be a builder implementation detail and an orthogonal concern. However, if this is non controversial, I am not opposed to adding it in. Would you mind to propose this as a suggestion? Should probably refer to the `pack` support for it and maybe briefly touch upon potential drawbacks?

Reply:
```suggestion
We would replace the current `base`, `tiny` and `static` stacks with a single stack that includes all packages required by the buildpacks. Similarly, we would replace the current `base`, `tiny` and `static` builders with a single builder.

Container images have a layer limit and in the past, we have had issues with builders that include a lot of buildpacks because this can quickly cause the builder to hit the layer limit. To prevent this from happening, we will by default squash layers for all composite buildpacks included in the builder. This will be done by using the `--flatten` of `pack builder create`, and it should allow the builder to scale to include all of the Paketo buildpacks.
```

I don't think this is controversial. The functionality exists and we've talked about doing it with our builders before, but I just don't think it was ever implemented.

As far as drawbacks, I don't think there are any major ones. You could potentially see larger image sizes because if you have individual layers per-buildpack and if you have a builder that has multiple copies of a buildpack, say the environment-variables buildpack, then you'd only get one layer for that buildpack. If you squash them so there's a layer per composite, then you won't get that layer sharing.

I don't think this will really matter because in reality, we're still getting duplicates because composites often have different versions of a buildpack and each version is its own layer. Also, having fewer layers as a whole is typically considered better for images (look at any major image on Docker Hub) and it will typically download faster because the client has to do more work for lots of little layers.